### PR TITLE
Adding parallel_each functions. Adding the execution_space as an argument for the for_each/parallel_each functions

### PR DESCRIPTION
**Pull request by thierryantoun**

### CODE REVIEW

**Review by tpadioleau — COMMENTED** [review]
I would advise to keep `ddc::for_each(dom, f)` and replace all previous `ddc::for_each(ddc::policies::parallel_device, dom, f);` by `ddc::parallel_for_each(dom, f);`

File: include/ddc/for_each.hpp [hunk 1]
```diff
@@ -164,62 +181,42 @@ inline void for_each(
     detail::for_each_serial<DiscreteElement<DDims...>>(begin, end, std::forward<Functor>(f));
 }
 
-/** iterates over a nD domain using the serial execution policy
+/** iterates over a nD domain using the default execution policy
```
Comment (tpadioleau) [inline include/ddc/for_each.hpp:184]:
```suggestion
/** iterates over a nD domain in serial
```

File: include/ddc/for_each.hpp [hunk 2]
```diff
@@ -164,62 +181,42 @@ inline void for_each(
...
-        DiscreteDomain<DDims...> const& domain,
-        Functor&& f) noexcept
+inline void for_each(DiscreteDomain<DDims...> const& domain, Functor&& f) noexcept
 {
-    detail::for_each_kokkos<Kokkos::DefaultHostExecutionSpace>(domain, std::forward<Functor>(f));
+    for_each(ddc::policies::serial_host, domain, std::forward<Functor>(f));
 }
 
-/** iterates over a nD domain using the parallel_device_policy execution policy
+/** iterates over a nD domain using the serial execution policy
```
Comment (tpadioleau) [inline include/ddc/for_each.hpp:194]:
```suggestion
/** iterates over a nD domain using a given `Kokkos` execution space
```

File: include/ddc/for_each.hpp [hunk 3]
```diff
@@ -164,62 +181,42 @@ inline void for_each(
...
-        parallel_device_policy,
+template <class ExecSpace, class... DDims, class Functor>
+inline void parallel_for_each(
+        ExecSpace&& execution_space,
         DiscreteDomain<DDims...> const& domain,
         Functor&& f) noexcept
 {
-    detail::for_each_kokkos<Kokkos::DefaultExecutionSpace>(domain, std::forward<Functor>(f));
+    detail::for_each_kokkos<
+            ExecSpace>(std::forward<ExecSpace>(execution_space), domain, std::forward<Functor>(f));
```
Comment (tpadioleau) [inline include/ddc/for_each.hpp:205]:
```suggestion
    detail::for_each_kokkos(std::forward<ExecSpace>(execution_space), domain, std::forward<Functor>(f));
```

File: include/ddc/for_each.hpp [hunk 4]
```diff
@@ -164,62 +181,42 @@ inline void for_each(
...
 }
 
 /** iterates over a nD domain using the default execution policy
  * @param[in] domain the domain over which to iterate
  * @param[in] f      a functor taking an index as parameter
  */
```
Comment (tpadioleau) [inline include/ddc/for_each.hpp:211]:
```suggestion
/** iterates over a nD domain using the `Kokkos` default execution space
 * @param[in] domain the domain over which to iterate
 * @param[in] f      a functor taking an index as parameter
 */
```

File: include/ddc/for_each.hpp [hunk 5]
```diff
@@ -164,62 +181,42 @@ inline void for_each(
...
-        ExecutionPolicy&& policy,
-        ChunkSpan<ElementType, DiscreteDomain<DDims...>, LayoutPolicy> chunk_span,
-        Functor&& f) noexcept
+inline void parallel_for_each(DiscreteDomain<DDims...> const& domain, Functor&& f) noexcept
 {
-    for_each(std::forward<ExecutionPolicy>(policy), chunk_span.domain(), std::forward<Functor>(f));
+    parallel_for_each<Kokkos::DefaultExecutionSpace>(
+            Kokkos::DefaultExecutionSpace(),
+            domain,
+            std::forward<Functor>(f));
```
Comment (tpadioleau) [inline include/ddc/for_each.hpp:218]:
```suggestion
    parallel_for_each(
            Kokkos::DefaultExecutionSpace(),
            domain,
            std::forward<Functor>(f));
```

File: include/ddc/kernels/splines/spline_builder.hpp [hunk 6]
```diff
@@ -559,68 +559,122 @@ operator()(
...
+                    exec_space(),
+                    batch_domain(),
+                    KOKKOS_LAMBDA(typename batch_domain_type::discrete_element_type j) {
+                        for (int i = s_nbc_xmin; i > 0; --i) {
+                            spline(ddc::DiscreteElement<bsplines_type>(s_nbc_xmin - i), j)
+                                    = derivs_xmin_values(ddc::DiscreteElement<deriv_type>(i), j)
+                                      * ddc::detail::ipow(dx_proxy, i + s_odd - 1);
+                        }
+                    });
+        }
```
Comment (tpadioleau) [inline include/ddc/kernels/splines/spline_builder.hpp:583]: You can also apply this to all the rest
```suggestion
            ddc::parallel_for_each(
                    exec_space(),
                    batch_domain(),
                    KOKKOS_LAMBDA(typename batch_domain_type::discrete_element_type j) {
                        for (int i = s_nbc_xmin; i > 0; --i) {
                            spline(ddc::DiscreteElement<bsplines_type>(s_nbc_xmin - i), j)
                                    = derivs_xmin_values(ddc::DiscreteElement<deriv_type>(i), j)
                                      * ddc::detail::ipow(dx_proxy, i + s_odd - 1);
                        }
                    });
```

File: tests/for_each.cpp [hunk 7]
```diff
@@ -73,47 +73,55 @@ TEST(ForEachSerialHost, TwoDimensions)
...
+TEST(ParallelForEachParallelHost, ZeroDimension)
 {
     DDom0D const dom;
     std::vector<int> storage(dom.size(), 0);
     ddc::ChunkSpan<int, DDom0D> const view(storage.data(), dom);
-    ddc::for_each(ddc::policies::parallel_host, dom, [=](DElem0D const i) { view(i) += 1; });
+    ddc::parallel_for_each<Kokkos::DefaultHostExecutionSpace>(
+            Kokkos::DefaultHostExecutionSpace(),
+            dom,
+            [=](DElem0D const i) { view(i) += 1; });
```
Comment (tpadioleau) [inline tests/for_each.cpp:88]:
```suggestion
    ddc::parallel_for_each(
            Kokkos::DefaultHostExecutionSpace(),
            dom,
            [=](DElem0D const i) { view(i) += 1; });
```

File: tests/for_each.cpp [hunk 8]
```diff
@@ -73,47 +73,55 @@ TEST(ForEachSerialHost, TwoDimensions)
...
+TEST(ParallelForEachParallelHost, OneDimension)
 {
     DDomX const dom(lbound_x, nelems_x);
     std::vector<int> storage(dom.size(), 0);
     ddc::ChunkSpan<int, DDomX> const view(storage.data(), dom);
-    ddc::for_each(ddc::policies::parallel_host, dom, [=](DElemX const ix) { view(ix) += 1; });
+    ddc::parallel_for_each<Kokkos::DefaultHostExecutionSpace>(
+            Kokkos::DefaultHostExecutionSpace(),
+            dom,
+            [=](DElemX const ix) { view(ix) += 1; });
```
Comment (tpadioleau) [inline tests/for_each.cpp:100]:
```suggestion
    ddc::parallel_for_each(
            Kokkos::DefaultHostExecutionSpace(),
            dom,
            [=](DElemX const ix) { view(ix) += 1; });
```

File: tests/for_each.cpp [hunk 9]
```diff
@@ -73,47 +73,55 @@ TEST(ForEachSerialHost, TwoDimensions)
...
+TEST(ParallelForEachParallelHost, TwoDimensions)
 {
     DDomXY const dom(lbound_x_y, nelems_x_y);
     std::vector<int> storage(dom.size(), 0);
     ddc::ChunkSpan<int, DDomXY> const view(storage.data(), dom);
-    ddc::for_each(ddc::policies::parallel_host, dom, [=](DElemXY const ixy) { view(ixy) += 1; });
+    ddc::parallel_for_each<Kokkos::DefaultHostExecutionSpace>(
+            Kokkos::DefaultHostExecutionSpace(),
+            dom,
+            [=](DElemXY const ixy) { view(ixy) += 1; });
```
Comment (tpadioleau) [inline tests/for_each.cpp:112]:
```suggestion
    ddc::parallel_for_each(
            Kokkos::DefaultHostExecutionSpace(),
            dom,
            [=](DElemXY const ixy) { view(ixy) += 1; });
```

**Review by tpadioleau — CHANGES_REQUESTED** [review]
Almost ready, no need to have a `if constexpr`, if `ExecSpace` is `Kokkos::Serial` then the `parallel_for_each` will handle it fine.

**Comment by thierryantoun** [comment]
But if the **parallel_for_each** function works fine with **Kokkos::Serial** as an execution space, why is the simple **for_each** function still needed?

**Comment by tpadioleau** [comment]
> But if the **parallel_for_each** function works fine with **Kokkos::Serial** as an execution space, why is the simple **for_each** function still needed?

One cannot call `Kokkos::parallel_for` with `Kokkos::Serial` inside a `Kokkos::parallel_for` with `Kokkos::Cuda`, we plan to support that with `ddc::for_each`. An other reason is that, `Kokkos::Serial` may not always be available depending on the compilation options. Only `Kokkos::DefaultHostExecutionSpace` and `Kokkos::DefaultExecutionSpace` are always available. This is linked to #225.

**Review by tpadioleau — APPROVED** [review]

**Pull request merged**